connectivity/tests/multicast.go

```diff
@@ -0,0 +1,363 @@
+		}
+	}
+	return errRet
+}
+
+// create multicast group and add all cilium nodes to the multicast group for testing
+func (s *socatMulticast) addAllNodes(ctx context.Context, t *check.Test) error {
+	ct := t.Context()
+	client := ct.K8sClient()
+
+	ciliumPodsList, err := client.ListPods(ctx, ct.Params().CiliumNamespace, metav1.ListOptions{LabelSelector: defaults.AgentPodSelector})
+	if err != nil {
+		return err
+	}
+	ciliumPods := ciliumPodsList.Items
+
+	// Create a map of ciliumInternalIPs of all nodes
+	ipToPodMap := make(map[v2.NodeAddress]string)
+	if err := s.populateMaps(ctx, t, ciliumPods, ipToPodMap); err != nil {
+		return err
+	}
+
+	var wg sync.WaitGroup
+	errCh := make(chan error, len(ciliumPods))
+	wg.Add(len(ciliumPods))
+
+	for _, ciliumPod := range ciliumPods {
+		go func(pod corev1.Pod) {
+			defer wg.Done()
+			// If there are not specified multicast group, create it
+			cmd := []string{"cilium-dbg", "bpf", "multicast", "subscriber", "list", testMulticastGroupIP}
+			_, stdErr, err := client.ExecInPodWithStderr(ctx, pod.Namespace, pod.Name, defaults.AgentContainerName, cmd)
+			if err != nil {
+				if !strings.Contains(stdErr.String(), notExistMsg) {
+					errMsg := fmt.Sprintf("Error: %v, Stderr: %s", err, stdErr.String())
+					errCh <- errors.New(errMsg)
+					t.Fatalf("Fatal error occurred while checking multicast group %s in %s", testMulticastGroupIP, pod.Spec.NodeName)
+					return
+				}
+				s.addNodeWithoutGroup(pod.Spec.NodeName)
+				cmd = []string{"cilium-dbg", "bpf", "multicast", "group", "add", testMulticastGroupIP}
+				_, stdErr, err := client.ExecInPodWithStderr(ctx, pod.Namespace, pod.Name, defaults.AgentContainerName, cmd)
+				if err != nil {
+					errMsg := fmt.Sprintf("Error: %v, Stderr: %s", err, stdErr.String())
+					errCh <- errors.New(errMsg)
+					t.Fatalf("Fatal error occurred while creating multicast group %s in %s", testMulticastGroupIP, pod.Spec.NodeName)
+					return
+				}
+			}
+			// Add all ciliumInternalIPs of all nodes to the multicast group as subscribers
+			for ip, podName := range ipToPodMap {
+				if ip.IP != "" && pod.Name != podName { // My node itself does not need to be in a multicast group.
+					cmd = []string{"cilium-dbg", "bpf", "multicast", "subscriber", "add", testMulticastGroupIP, ip.IP}
+					_, stdErr, err := client.ExecInPodWithStderr(ctx, pod.Namespace, pod.Name, defaults.AgentContainerName, cmd)
+					if err == nil {
+						s.addNotSubscribePodAddress(pod.Spec.NodeName, ip)
+					} else if !strings.Contains(stdErr.String(), alreadyExistMsg) {
+						errMsg := fmt.Sprintf("Error: %v, Stderr: %s", err, stdErr.String())
+						errCh <- errors.New(errMsg)
+						t.Fatalf("Fatal error occurred while adding node %s to multicast group %s in %s", ip.IP, testMulticastGroupIP, pod.Spec.NodeName)
+						return
+					}
+				}
+			}
+		}(ciliumPod)
+	}
+
+	wg.Wait()
+	close(errCh)
+
+	var errRet error
+	for fetchdata := range errCh {
+		if fetchdata != nil {
```

Comment:
[errors.Join discards nil values](https://pkg.go.dev/errors#Join), so you likely don't need this nil check here.